lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3474,38 +3489,48 @@ void replaceScalarArrayWithVectorArray(Value *ScalarArray, Value *VectorArray,
     if (U->user_empty())
       continue;
     if (BitCastInst *BCI = dyn_cast<BitCastInst>(U)) {
-      BCI->setOperand(0, VectorArray);
+      // Avoid replacing the dest of the memcpy to support partial replacement.
+      if (MC->getArgOperand(0) != BCI)
+        BCI->setOperand(0, VectorArray);
```

Comment:
AFAK, the dest of the memcpy inst will always be a bitcast instruction (bitcast to i8*), since that's the way it generates the memcpy.  I believe we have a lot of other code here that depends on that assumption.  I don't think there ever should be an addrspacecast between the dest pointer and the memcpy.